### PR TITLE
Fix race condition when rapidly starting/stopping health check

### PR DESCRIPTION
**Pull request by choo-stripe**
## Description
There's an interesting race condition when starting/stopping a healthcheck.
1. Create `NewHealthCheck`
2. `NewHealthCheck` spawns topo-watchers in goroutines
3. `hc.Close()` is called
4. `Stop()` is called on the topo-watchers
5. The topo-watcher cancels the ctx and calls `Wait()` on the wait-group. The `Wait()` returns immediately, because the `Start()` goroutine hasn't even started yet. So nothing was ever added to the wait-group.
6. You close the topo-srv, which should be safe, because the health-check has already been closed
7. The `Start()` goroutine now fires, which calls `tw.loadTablets()` (before even checking ctx!). This will NPE because the topo-srv has already been closed.

I actually managed to reproduce this with a test.

There's another potential fix here:
- Move `loadTablets` under `case <-ticker.C`. Note that [`cancelFunc` is called before `Wait()`](https://github.com/vitessio/vitess/blob/f085f12b1cd38a9bcf907604a747af308e93c4fd/go/vt/discovery/topology_watcher.go#L142-L146). When `Stop()` is called, if we've already added to the wait-group, than `wg.Wait()` will make sure the goroutine actually exits. If we _haven't_ yet added to the wait-group, `tw.ctx.Done()` will fire before `ticker.C`, and again, we won't load tablets.

The last, unrelated fix I made is to check if the cell to watch is `""`. Splitting empty string cannot yield length 0: https://play.golang.org/p/5mKKN9TkMh5, so that check was not doing anything.

Lastly, I validated there were no other call-sites of `Start()` that we'd have to get rid of `go` in front:
```
rg 'go .*\.Start()'
```

## Related Issue(s)
- N/A

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

## Deployment Notes
N/A

### CODE REVIEW

File: go/vt/discovery/healthcheck_test.go [hunk 1]
```diff
@@ -505,6 +505,16 @@ func TestHealthCheckCloseWaitsForGoRoutines(t *testing.T) {
...
 }
 
+// TestHealthCheckStartCloseRace tests that rapidly calling Start()/Close()
+// still cleanly shuts down.
+func TestHealthCheckStartCloseRace(t *testing.T) {
+	ts := memorytopo.NewServer("cell")
+	hc := createTestHc(ts)
+	assert.Nil(t, hc.Close(), "Close returned error")
+	ts.Close()
+	time.Sleep(50 * time.Millisecond)
```
Comment (systay) [inline go/vt/discovery/healthcheck_test.go:515]: I don't get it. why sleep at the end of the method like this?

Comment (choo-stripe) [reply go/vt/discovery/healthcheck_test.go:515]: Without it, the race/NPE was not triggered. Basically, we need to give the health-check time to actually run the goroutine on a closed topo-server in order to get an NPE. This test-case is for a super-niche bug, so I'd be fine removing if you want.

Comment (frouioui) [reply go/vt/discovery/healthcheck_test.go:515]: Is there a way we can add an assertion to assert the health check is correctly shut down on line 515?

Comment (frouioui) [reply go/vt/discovery/healthcheck_test.go:515]: note that the `time.Sleep` might be a bit sensitive as it can introduce flakiness in the CI builds

**Review by frouioui — APPROVED** [review]
I left a comment to add, if it is possible, an assertion in the test you've added. Otherwise, it looks good to me.

**Review by deepthi — COMMENTED** [review]
Nice fix! The code change itself looks good to me. Can you do the following?
- fix DCO, we cannot merge until that is fixed
- address the other feedback
- squash the commits when you fix the DCO (that will anyway require a force-push)

**Comment by choo-stripe** [comment]
@frouioui I actually couldn't arrive at a good strategy to test this. Thoughts:
- It doesn't make sense to test this from the healthchecker, because it actually did nothing wrong. It was correctly calling `Stop` on the topo-watcher
- So I tried to create a test for the topo-watcher. But! There's a potential race in tests. If we:
    - create topo-watcher
    - `tw.Start()`
    - `tw.Stop()`
    The topo-watcher may or may not have already refreshed tablets. So we can't e.g. check for tablet refresh, because both a refresh and no refresh are acceptable. Also! Before the fix, we called `go tw.Start()`. We have no way to know if the topo-watcher would have added to the wait-group or not by the time we `Stop()`. So we can't check that. And we don't get signaled when the topo-watcher goroutines exit (that was the bug, it didn't exit). So without mucking around with the topo-watcher struct and adding test-only status flags, I don't have good ideas to validate that `Stop()` actually closes the goroutines.
- With the changes in this pr, we could create a simple test:
    ```
    // create topo-watcher
    tw.Start()
    tw.Stop()
    tw.wg.Wait()
    ```
    This would pass, and verify that all the `wg` goroutines actually exit. But it wouldn't fail with the old implementation, it would hang because `tw.Start()` is blocking. I'm having trouble coming up with a test that fails in the old implementation, and now passes.

Any ideas?

I did sign off on the commit and rebase. I also deleted the test.

File: go/vt/vtgate/tabletgateway_test.go [hunk 2]
```diff
@@ -104,7 +104,8 @@ func TestTabletGatewayBeginExecuteBatch(t *testing.T) {
 }
 
 func TestTabletGatewayShuffleTablets(t *testing.T) {
-	tg := NewTabletGateway(context.Background(), nil, nil, "local")
+	hc := discovery.NewFakeHealthCheck(nil)
+	tg := NewTabletGateway(context.Background(), hc, nil, "local")
```
Comment (deepthi) [inline go/vt/vtgate/tabletgateway_test.go:108]: Is this change really required?

**Comment by deepthi** [comment]
@frouioui what are your thoughts on testing?

**Comment by frouioui** [comment]
With this new implementation, where the main thread is blocked until all the topo watchers are added to the WaitGroup, I find it hard to create a test to assert a possible race condition. I think it is okay to leave it as it is now, aka with no additional tests. The change looks safe enough to me.

**Comment by frouioui** [comment]
Hello @choo-stripe and @deepthi, I took the liberty to add a test case. The test ensures that rapidly starting and stopping the topo watcher works as intended.

File: go/vt/discovery/topology_watcher_test.go [hunk 3]
```diff
@@ -59,6 +59,52 @@ func checkChecksum(t *testing.T, tw *TopologyWatcher, want uint32) {
...
+	tw.Stop()
+	done <- true
+
+	tw.wg.Wait()
+	done <- true
+
+	_, ok := <-result
+	if !ok {
+		t.Fatal("timed out")
+	}
```
Comment (deepthi) [inline go/vt/discovery/topology_watcher_test.go:105]: nit: we should use `require` instead of `t.Fatal`.
The rest LGTM.

**Review by deepthi — APPROVED** [review]
